mm2src/coins/eth/eth_swap_v2/eth_taker_swap_v2.rs

```diff
@@ -457,37 +458,120 @@ impl EthCoin {
+        let decoded = decode_contract_call(approve_func, taker_payment.unsigned().data())?;
+        let id = match decoded.first() {
+            Some(Token::FixedBytes(bytes)) => bytes,
+            invalid_token => {
+                return MmError::err(FindPaymentSpendError::InvalidData(format!(
+                    "Expected Token::FixedBytes, got {:?}",
+                    invalid_token
+                )))
+            },
+        };
+        // loop to find maker's spendTakerPayment transaction
         loop {
-            let taker_status = self
-                .payment_status_v2(
-                    taker_swap_v2_contract,
-                    decoded[0].clone(), // id from spendTakerPayment
-                    &TAKER_SWAP_V2,
-                    EthPaymentType::TakerPayments,
-                    TAKER_PAYMENT_STATE_INDEX,
-                )
-                .await?;
-            if taker_status == U256::from(TakerPaymentStateV2::MakerSpent as u8) {
-                return Ok(taker_payment.clone());
-            }
             let now = now_sec();
             if now > wait_until {
-                return MmError::err(WaitForPaymentSpendError::Timeout { wait_until, now });
+                return MmError::err(FindPaymentSpendError::Timeout { wait_until, now });
             }
-            Timer::sleep(10.).await;
+
+            let current_block = match self.current_block().compat().await {
+                Ok(b) => b,
+                Err(e) => {
+                    error!("Error getting block number: {}", e);
+                    Timer::sleep(5.).await;
+                    continue;
+                },
+            };
+
+            // get all logged TakerPaymentSpent events from `from_block` till current block
+            let events = match self
+                .events_from_block(taker_swap_v2_contract, "TakerPaymentSpent", from_block, current_block)
+                .await
+            {
+                Ok(events) => events,
+                Err(e) => {
+                    error!(
+                        "Error getting TakerPaymentSpent events from {} block: {}",
+                        from_block, e
+                    );
+                    Timer::sleep(5.).await;
+                    continue;
+                },
+            };
+
+            if events.is_empty() {
+                info!(
+                    "No events found yet for the block range {} to {}",
+                    from_block, current_block
+                );
+                Timer::sleep(5.).await;
+                continue;
+            }
+
+            // this is how spent event looks like in EtomicSwapTakerV2: event TakerPaymentSpent(bytes32 id, bytes32 secret)
+            let found_event = events.into_iter().find(|event| &event.data.0[..32] == id.as_slice());
+
+            if let Some(event) = found_event {
+                if let Some(tx_hash) = event.transaction_hash {
```

Comment:
I think this code could be simplified a bit, like this:
```
                if let Some(tx_hash) = event.transaction_hash {
                    match self.transaction(TransactionId::Hash(tx_hash)).await {
                        Ok(Some(transaction)) => {
                            return Ok(signed_tx_from_web3_tx(transaction).map_err(FindPaymentSpendError::Internal)?)
                        },
                        Ok(None) => info!("Tx {} not found yet", tx_hash),
                        Err(e) => error!("Get tx {} error: {}", tx_hash, e),
                    };
                    Timer::sleep(check_every).await;
                    continue;
                }
```

Reply:
Done https://github.com/KomodoPlatform/komodo-defi-framework/pull/2261/commits/1e7a197a4616c74a52f610583656bf12b179a326

Only one note, I prefer to create a variable first to return it, if we use quite long function. 
Its difficult to read function and mapping error in the Ok
`return Ok(signed_tx_from_web3_tx(transaction).map_err(FindPaymentSpendError::Internal)?)`